ocaml/testsuite/tests/typing-layouts/allow-illegal-crossing.ml

```diff
@@ -0,0 +1,654 @@
+and b = private a
+|}]
+
+module rec A : sig
+  type t : value mod portable uncontended
+end = struct
+  type t : value mod portable uncontended = private B.t
+end
+and B : sig
+  type t : value
+end = struct
+  type t : value
+end
+[%%expect {|
+module rec A : sig type t : value mod portable uncontended end
+and B : sig type t : value end
+|}]
+
+module rec A : sig
+  type t : value
+end = struct
+  type t : value
+end
+and B : sig
+  type t : value mod portable uncontended
+end = struct
+  type t : value mod portable uncontended = private A.t
+end
+[%%expect {|
+module rec A : sig type t : value end
+and B : sig type t : value mod portable uncontended end
+|}]
+
+type t : value mod portable = private { foo : string }
+[%%expect {|
+type t = private { foo : string; }
+|}]
+
+type a : value mod portable = { foo : string }
+type b : value mod portable = a = { foo : string }
+[%%expect {|
+type a = { foo : string; }
+type b = a = { foo : string; }
+|}]
+
+type a : value mod uncontended = private { foo : string }
+type b : value mod uncontended = a = private { foo : string }
+[%%expect {|
+type a = private { foo : string; }
+type b = a = private { foo : string; }
+|}]
+
+type t : value mod uncontended = private Foo of int | Bar
+[%%expect {|
+type t = private Foo of int | Bar
+|}]
+
+type a : value mod uncontended = Foo of int | Bar
+type b : value mod uncontended = a = Foo of int | Bar
+[%%expect {|
+type a = Foo of int | Bar
+type b = a = Foo of int | Bar
+|}]
+
+type a : value mod portable = private Foo of int | Bar
+type b : value mod portable = a = private Foo of int | Bar
+[%%expect {|
+type a = private Foo of int | Bar
+type b = a = private Foo of int | Bar
+|}]
+
+(********************************************)
+(* Test 2: illegal mode crossing propogates *)
+
+type a : value mod portable uncontended = private string
+type ('a : value mod portable uncontended) b
+type c = a b
+[%%expect {|
+type a = private string
+type ('a : value mod uncontended portable) b
+type c = a b
+|}]
+
+type t : value mod portable uncontended = private string
+let f : ('a : value mod portable uncontended). 'a -> 'a = fun x -> x
+let g (x : t) = f x
+[%%expect {|
+type t = private string
+val f : ('a : value mod uncontended portable). 'a -> 'a = <fun>
+val g : t -> t = <fun>
+|}]
+
+type t : value mod portable uncontended = { a : int; b : int }
+let x : _ as (_ : value mod portable uncontended) = { a = 5; b = 5 }
+[%%expect {|
+type t = { a : int; b : int; }
+val x : t = {a = 5; b = 5}
+|}]
+
+type ('a, 'b) t : value mod portable uncontended = { a : 'a; b : 'b }
+let x : _ as (_ : value mod portable uncontended) = { a = 5; b = 5 }
+[%%expect {|
+type ('a, 'b) t = { a : 'a; b : 'b; }
+val x : (int, int) t = {a = 5; b = 5}
+|}]
+
+type t : value mod portable uncontended = Foo of string | Bar of int
+let x : _ as (_ : value mod portable uncontended) = Foo "hello world"
+[%%expect {|
+type t = Foo of string | Bar of int
+val x : t = Foo "hello world"
+|}]
+
```

Comment:
I would want to see validation of the testing technique here. Something like

```
let x : _ as (_ : value mod portable uncontended) = "hi"
```

which would be rejected (and would show that actually this property is being correctly checked).